src/pool/srv_target.c

```diff
@@ -1094,113 +1090,171 @@ ds_pool_put(struct ds_pool *pool)
 	}
 	ABT_mutex_unlock(pool->sp_mutex);
 
-	if (pool->sp_stopping) {
-		D_DEBUG(DB_MD, DF_UUID": skip fetching hdl due to stop\n",
-			DP_UUID(pool->sp_uuid));
-		D_GOTO(out, rc);
-	}
-	D_INFO(DF_UUID": fetching handles\n", DP_UUID(pool->sp_uuid));
-	rc = ds_pool_iv_conn_hdl_fetch(pool);
-	if (rc) {
-		D_INFO(DF_UUID" iv conn fetch %d\n", DP_UUID(pool->sp_uuid), rc);
-		D_GOTO(out, rc);
-	}
+	while (!eph_report_exiting(pool)) {
+		sleep_intvl = EPH_REPORT_INTVL;
+
+		/* Fetch pool connection handles */
+		if (!conn_hdl_fetched) {
+			D_INFO(DF_UUID ": Fetching connection handles.\n", DP_UUID(pool->sp_uuid));
+			rc = ds_pool_iv_conn_hdl_fetch(pool);
+			if (rc) {
+				D_INFO(DF_UUID ": Failed to fetch connection handles. " DF_RC "",
+				       DP_UUID(pool->sp_uuid), DP_RC(rc));
+				sleep_intvl = EPH_REPORT_RETRY_INTVL;
+			} else {
+				conn_hdl_fetched = true;
+			}
+		}
 
-out:
-	D_INFO(DF_UUID": signaling done\n", DP_UUID(pool->sp_uuid));
-	ABT_mutex_lock(pool->sp_mutex);
-	ABT_cond_signal(pool->sp_fetch_hdls_done_cond);
-	ABT_mutex_unlock(pool->sp_mutex);
+		if (eph_report_exiting(pool))
+			break;
 
-	pool->sp_fetch_hdls = 0;
-	D_INFO(DF_UUID": end\n", DP_UUID(pool->sp_uuid));
-}
+		/* Fetch server pool/cont open handles */
+		if (!srv_hdl_fetched) {
+			D_INFO(DF_UUID ": Fetching srv open handles.\n", DP_UUID(pool->sp_uuid));
+			rc = ds_pool_iv_srv_hdl_fetch(pool, &pool->sp_srv_pool_hdl,
+						      &pool->sp_srv_cont_hdl);
+			if (rc) {
+				DL_ERROR(rc, DF_UUID ": Failed to fetch srv open handles.",
+					 DP_UUID(pool->sp_uuid));
+				sleep_intvl = EPH_REPORT_RETRY_INTVL;
+			} else if (uuid_is_null(pool->sp_srv_pool_hdl) ||
+				   uuid_is_null(pool->sp_srv_cont_hdl)) {
+				/* Feteched NULL server handle ?! */
+				D_INFO(DF_UUID ": Got NULL handle, ph=" DF_UUID ", coh=" DF_UUID
+					       "\n",
+				       DP_UUID(pool->sp_uuid), DP_UUID(pool->sp_srv_pool_hdl),
+				       DP_UUID(pool->sp_srv_cont_hdl));
+				sleep_intvl = EPH_REPORT_RETRY_INTVL;
+			} else {
+				rc = ds_pool_srv_open(pool->sp_uuid, pool->sp_srv_cont_hdl);
+				if (rc) {
+					DL_ERROR(rc, DF_UUID ": Failed to open srv cont hdl.",
+						 DP_UUID(pool->sp_uuid));
+				} else {
+					srv_hdl_fetched = true;
+				}
+			}
+		}
 
-static void
-tgt_track_eph_query_ult(void *data)
-{
-	ds_cont_track_eph_query_ult(data);
+		if (eph_report_exiting(pool))
+			break;
+
+		/* Report EC agg epoch boundary */
+		rc = ds_cont_eph_report(pool);
+		if (rc) {
+			DL_ERROR(rc, "Failed to report EC agg epoch.");
+			sleep_intvl = EPH_REPORT_RETRY_INTVL;
+		}
+
+		if (eph_report_exiting(pool))
+			break;
+
+		sched_req_sleep(pool->sp_ec_ephs_req, sleep_intvl);
+	}
+out:
+	D_DEBUG(DB_MD, DF_UUID " Exit eph report ULT.\n", DP_UUID(pool->sp_uuid));
 }
 
 static int
-ds_pool_start_track_eph_query_ult(struct ds_pool *pool)
+start_eph_report_ult(struct ds_pool *pool)
 {
 	struct sched_req_attr	attr;
 	uuid_t			anonym_uuid;
 
-	if (unlikely(ec_agg_disabled))
-		return 0;
```

Comment:
Because your patch combines the fetching connection handle & `srv_cont_hdl` together with `agg_eph_report` logic together, then even if `ec_agg_disabled`, we still have to start `eph_report_ult` ult, that seems some confused. Maybe we can add some check inside `eph_report_ult` or elsewhere?

Reply:
'ec_agg_disabled' is for debug only, and it looks to me it's not quite related with starting the epoch reporting ULT, so I'd keep it simple.